common/changes/@cadl-lang/rest/metadata4_2022-09-16-18-45.json

```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@cadl-lang/rest",
+      "comment": "Implement automatic visibility transformations",
```

Comment:
@daviwil Moving this here to get a thread going.


> I can tell this must have been a major headache to write! 

I considered career change instead, lol.


> Just to verify, the visibility filtering works recursively too? I see some tests that appear to verify that but I wasn't 100% sure.

Yes, it is recursive. Which is both why it`s hard and also the most needle moving thing vs. explicit visibility handling.

> How much harder do you think this will be to maintain going forward? The implementation is obviously fine, but it does a lot now

My earlier draft would have been a nightmare to maintain which is why I held it back for so long. But yes, that is more about implementation than the intrinsic issues with the approach you raise. But this implementation is 10x simpler FWIW, so that`s good at least.

> I wonder what kind of cascading effects we'll get when we add a new property with a visibility setting that causes a schema to be split into multiple visibility-filtered schemas?

The mitigation here is that another emitter can decide not to split and I don't think client emitters will split. In fact, I see a lot of questions about how do I recover the original type from client emitters that see the impacts of the resource interfaces transforming the types via templates, etc.

Fundamentally, though, there isn`t really any way to express this in OpenAPI without generating schemas. You can`t even inline in the general case to at least avoid naming split off single-purpose schemas because you can`t inline cycles.

> I wonder what kind of freaky effects we'll get when a new property is `@added` and has a visibility that would cause a schema to be split. Have you tried that?

Hmm, great point, maybe this can introduce the possibility of a breaking change with `@added` which would be very bad. Maybe we can detect this and error, but that also sounds very complicated. :(

Reply:
I will file an issue to look at this.

Reply:
https://github.com/microsoft/cadl/issues/1072